clippy_lints/src/methods/unnecessary_fallible_conversions.rs

```diff
@@ -2,25 +2,74 @@ use clippy_utils::diagnostics::span_lint_and_then;
 use clippy_utils::get_parent_expr;
 use clippy_utils::ty::implements_trait;
 use rustc_errors::Applicability;
-use rustc_hir::{Expr, ExprKind};
+use rustc_hir::{Expr, ExprKind, QPath};
 use rustc_lint::LateContext;
 use rustc_middle::ty;
 use rustc_middle::ty::print::with_forced_trimmed_paths;
 use rustc_span::{sym, Span};
 
 use super::UNNECESSARY_FALLIBLE_CONVERSIONS;
 
+#[derive(Copy, Clone)]
+enum SpansKind {
+    TraitFn { trait_span: Span, fn_span: Span },
+    Fn { fn_span: Span },
+}
+
 /// What function is being called and whether that call is written as a method call or a function
 /// call
 #[derive(Copy, Clone)]
 #[expect(clippy::enum_variant_names)]
 enum FunctionKind {
     /// `T::try_from(U)`
-    TryFromFunction,
+    TryFromFunction(Option<SpansKind>),
     /// `t.try_into()`
     TryIntoMethod,
     /// `U::try_into(t)`
-    TryIntoFunction,
+    TryIntoFunction(Option<SpansKind>),
+}
+
+impl FunctionKind {
+    fn applicability(&self, parent_unwrap_call: &Option<Span>) -> Applicability {
```

Comment:
Hmm one last thing. I find it a bit weird to do this over `fn sugg()` and return either `default_sugg`/`machine_applicable_sugg` plus the applicability.

It just feels, I dunno, a bit error prone to do `== MachineApplicable` 😅 and weird as usually applicability is lowered to `HasPlaceholders` if we fail to get a snippet of a span.

(Return type will be pretty complex tho. I think this is a sign we need some type alias for a multipart suggestion in the future...)

r=me after this

Reply:
PS the code is fine - I just think it should be rearranged a bit so that we don't match on `Applicability`. Having a fallback like this for when it specifically isn't machine applicable is oke

Reply:
Merging the calculation of `sugg` and `applicability` into the same process would indeed make it more concise, then it would become:

```rust
fn sugg_appl(&self, parent_unwrap_call: &Option<Span>, primary_span: Span) -> (Applicability, Vec<(Span, String)>) {
        let Some(unwrap_span) = parent_unwrap_call else {
            return ...
        };

        match &self {
            FunctionKind::TryFromFunction(None) | FunctionKind::TryIntoFunction(None) => ...,
            _ => ...,
        }
    }
```

Do I need to make such a change?

Reply:
Yea smth like that.

> Do I need to make such a change?

I'd prefer so. tho I can merge it now and create a followup PR tomorrow once I have my desktop back